File: src/main/java/com/baulsupp/oksocial/credentials/ServiceDefinition.java
package com.baulsupp.oksocial.credentials;

import java.util.Optional;

import static java.util.Optional.empty;

public interface ServiceDefinition<T> {
  String apiHost();

  String serviceName();

  T parseCredentialsString(String s);

  String formatCredentialsString(T credentials);

  String shortName();

  default Optional<String> apiDocs() {
    return empty();
  }

  default Optional<String> accountsLink() {
    return empty();
  }
}

File: src/main/java/com/baulsupp/oksocial/credentials/ServiceDefinition.kt
package com.baulsupp.oksocial.credentials

import java.util.Optional

import java.util.Optional.empty

interface ServiceDefinition<T> {
    fun apiHost(): String

    fun serviceName(): String

    fun parseCredentialsString(s: String): T

    fun formatCredentialsString(credentials: T): String

    fun shortName(): String

    open fun apiDocs(): Optional<String> {
        return empty()
    }

    open fun accountsLink(): Optional<String> {
        return empty()
    }
}
